### PR TITLE
feat(purl): add generate purl package utilities

### PR DESCRIPTION
**Pull request by masahiro331**
Add generate purl utilities.  
specification: https://github.com/package-url/purl-spec

for: https://github.com/aquasecurity/trivy/pull/1522, https://github.com/aquasecurity/trivy/pull/1081

### CODE REVIEW

**Comment by masahiro331** [comment]
@AndreyLevchenko
I created a useful package to generate purl. Can you please review it?

File: pkg/purl/purl.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	"github.com/aquasecurity/fanal/analyzer"
+	"github.com/aquasecurity/fanal/analyzer/os"
+	"github.com/aquasecurity/fanal/types"
+
+	"github.com/package-url/packageurl-go"
+)
+
+func NewPackageURL(t string, pkg types.Package) packageurl.PackageURL {
+	name := strings.ReplaceAll(pkg.Name, ":", "/")
```
Comment (AndreyLevchenko) [inline pkg/purl/purl.go:15]: could you add test case for the replacement?

File: pkg/purl/purl.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	"strings"
+
+	"github.com/aquasecurity/fanal/analyzer"
+	"github.com/aquasecurity/fanal/analyzer/os"
+	"github.com/aquasecurity/fanal/types"
+
+	"github.com/package-url/packageurl-go"
+)
+
+func NewPackageURL(t string, pkg types.Package) packageurl.PackageURL {
```
[the remark this answers is not in the document]
Comment (masahiro331) [reply pkg/purl/purl.go:14]: What are the cases in which the Package structure cannot be used?

File: pkg/purl/purl.go (hunk 1 above)
Comment (masahiro331) [reply pkg/purl/purl.go:15]: thx. I fixed

File: pkg/purl/purl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (masahiro331) [reply pkg/purl/purl.go:14]: I understood. 
https://github.com/aquasecurity/trivy/pull/1522/files#diff-9bf6e24337d9bac10703d1bf42439dc126b97de4872e8b4a7460fad65b88ff23R108

Can't I just recreate the Package structure and put it in?
If we include the OS package, we will need many arguments.

[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply pkg/purl/purl.go:14]: could you add
```go
func NewPackageURL(t string, vln types.DetectedVulnerability) packageurl.PackageURL
```

[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply pkg/purl/purl.go:14]: ah actually it needs to be named differently

**Comment by knqyf263** [comment]
It should be aligned with the spec. See examples.
https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst

File: pkg/purl/purl.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+			Key:   "release",
+			Value: pkg.Release,
+		})
+	}
+	if pkg.Epoch != 0 {
+		qualifiers = append(qualifiers, packageurl.Qualifier{
+			Key:   "epoch",
+			Value: strconv.Itoa(pkg.Epoch),
+		})
+	}
```
Comment (knqyf263) [inline pkg/purl/purl.go:83]: They should be included in `version`.
https://github.com/package-url/purl-spec#purl

**Review by AndreyLevchenko — APPROVED** [review]

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**